texlive.combine: split static $TEXMFDIST into a separate derivation

xworld21 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
This is a tentative fix for #118745, with the least intrusive change I could think of. The behaviour should be as correct as it gets, since now the linking is done with two calls to `buildEnv`, instead of generating the `bin/` folder manually. ~The package man pages are still done by hand, unfortunately, but I have checked the tlpdb file, and it seems safe to do.~ Edit: the man pages are handled reliably now.

**Note.** The patches exposes a `texmf-dist` directory from the binary packages that was not there before, including yet another copy of `texmf.cnf` from `texlive-bin`. This should have no effect, as `TEXMFDIST` is already set to `texmf` anyway.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS (`scheme-full`)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented:
Please drop the unrelated packages.

SuperSandro2000 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -252,13 +250,23 @@ in (buildEnv {
-        ln -s -t . ../texmf/doc/"$d"/*
-      )
+    for t in {texmf,texmf-dist}; do
+      for d in {man/man1,man/man5}; do
```
```suggestion
      for d in man/man{1,5}; do
```

SuperSandro2000 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -252,13 +250,23 @@ in (buildEnv {
-        mkdir -p "./share/$d" && cd "./share/$d"
-        ln -s -t . ../texmf/doc/"$d"/*
-      )
+    for t in {texmf,texmf-dist}; do
```
```suggestion
    for t in texmf{,-dist}; do
```

xworld21 commented:
@SuperSandro2000 sorry about that, I switched to staging on the UI without knowing the consequences. I have rebased against staging.

veprbl commented:
I wonder, how is this different from `paths = pkgList.all`?

xworld21 commented:
> I wonder, how is this different from `paths = pkgList.all`?

We need the `extraPrefix = "/share/texmf"` line somewhere. Currently, the run files are unpacked directly into `$out`, so they need to be relocated before being combined with the rest.

I thought of (and implemented, actually) changing the unpack phase of the packages, but it was a lot of work and I didn't really see any benefit. It would even break texlive for the the users who combine their own packages.

Moving towards a "split" output model doesn't seem that bad either, if we could e.g. generate the `ls-R` directly into `-tex` we would uncouple the packages from the binaries. But I digress – I was really going for a minimal robust fix here.

xworld21 commented:
By the way, I forgot to say that I wasn't sure about staging vs master due to the rebuild – could you please make that decision for me (especially with 21.05 being fairly close now)?

veprbl commented:
@xworld21 So, you change `path` of the outer `buildEnv` to process `pkgList.bin`. This means that the symlinks for `pkgList.nonbin` will be processed and stored twice. If we are to introduce this inner `buildEnv`, it would make sense to have it process `pkgList.bin` that is a shorter list than `pkgList.nonbin`.

/rebase master

veprbl commented:
/rebase master

SuperSandro2000 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -38,16 +38,34 @@ let
+    extraPrefix = "/share/texmf";
+    ignoreCollisions = false;
+    paths = pkgList.nonbin;
+    postBuild =
+    # link info and man pages
+    ''
+      for d in {info,man}; do
+        [[ -e "$out/share/texmf/doc/$d" ]] && ln -s "texmf/doc/$d" "$out/share/$d"
+      done
+    '';
```
Can we replace this with ``outputsToInstall``?

xworld21 replied:
If I understand correctly (both the question and `texlive.combine`), not in the way the packages are generated. For this `buildEnv`, each package is bringing only one output.

xworld21 commented:
> it would make sense to have it process `pkgList.bin` that is a shorter list than `pkgList.nonbin`.

Can we actually do that? The outer `buildEnv` will have `extraPrefix`, so you get an odd `share/texmf/bin/` folder. Then we are back to linking the files manually, at which point, I'd rather extend the current `bin/` hack to the other folders. It's certainly quick to do.

OTOH, generating the `ls-R` files and the hypenation patterns within `texmf`, and *omitting* `texmf` from the final output, sounds a lot better. No duplicate links, not even in your profile.

I'll follow your lead: if this is better done fast, I'll try the quick fix. Otherwise I'll experiment with the `ls-R` approach.

PS. I can quantify the duplicate links with the full scheme: links to `share/*` folders (11), then to `scripts` (28), `tex/generic` (311), `tex/generic/config` (uh... some multiple of the number of languages). Technically `web2c` is not a link yet, but it is very easy to make it so.

xworld21 commented:
> This means that the symlinks for `pkgList.nonbin` will be processed and stored twice.

Ok, updated strategy: `texmf` generates its own `ls-R` database, and the configuration files (`texmf.cnf`, `languages.dat`, etc) pop up in `$TEXMFSYSVAR`. So now the duplicate links are `share/texmf` (just the one!) and the man pages, which is acceptable I hope?

xworld21 commented:
I have added a commit to do `mktexlsr --sort` in light of #102917. The `texmf` output should be reproducible now.

I compiled `scheme-full` and played around, and it seems to work just as well (or as bad) as before.

xworld21 commented:
@veprbl I have revived this PR. Same as my original attempt, I split off `$TEXMFDIST`, which contains the immutable files from TeX Live, from the files generated by combine (configuration, formats, etc).

The immediate advantage is that all man pages from `texlive.bin` will pop up in the combined output. This will temporarily make the clashes with doc packages from TeX Live worse. Future gain: I want to compile each format with its minimal TEXMFDIST (based on the triggers specified in the tlpdb), so that adding a package to texlive.combine becomes a quick process. Lots of other work before getting there, though.

xworld21 commented:
@GrahamcOfBorg build tests.texlive

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -132,18 +139,30 @@ in (buildEnv {
-      [[ -e "$fname" ]] && sed -E -n -f '${script}' -i "$fname"
+    mkdir -p "$TEXMFSYSVAR/tex/generic/config"
+    for fname in tex/generic/config/language.{dat,def}; do
+      [[ -e "$TEXMFDIST/$fname" ]] && sed -E -n -f '${script}' "$TEXMFDIST/$fname" > "$TEXMFSYSVAR/$fname"
```
Same pattern here. New directory, some files could be left behind?

xworld21 replied:
Again, Kpathsea will look up both `TEXMFSYSVAR` and `TEXMFDIST`, so we shouldn't worry about missed files.

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -132,18 +139,30 @@ in (buildEnv {
+    [[ -e "$TEXMFDIST"/web2c/fmtutil.cnf ]] && sed -E -f '${fmtutilSed}' "$TEXMFDIST"/web2c/fmtutil.cnf > "$TEXMFCNF"/fmtutil.cnf
+
+    # make new files visible to kpathsea
+    perl "$TEXMFDIST"/scripts/texlive/mktexlsr.pl --sort "$TEXMFSYSVAR"
```
Is this not done down below again?

xworld21 replied:
This needs to happen before we call `fmtutil`, so it must be here (let's just say, it took me a while to figure this one out...).

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -224,25 +242,26 @@ in (buildEnv {
     [[ -f "$TEXMFSYSCONFIG"/web2c/updmap.cfg ]] && sort -o "$TEXMFSYSCONFIG"/web2c/updmap.cfg "$TEXMFSYSCONFIG"/web2c/updmap.cfg
 
-    perl "$out"/share/texmf/scripts/texlive/mktexlsr.pl --sort "$out"/share/texmf-* # to make sure
+    perl "$TEXMFDIST"/scripts/texlive/mktexlsr.pl --sort "$TEXMFSYSCONFIG" "$TEXMFSYSVAR" # to make sure
```
(Here?)

xworld21 replied:
Rather, it is quite likely that this call is not necessary. I left it because it was already there, and there may have been a good reason (the comment "to be sure" is scarce on details, unfortunately!).

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -224,25 +242,26 @@ in (buildEnv {
       tName="$(basename $s | sed 's/\.[a-z]\+$//')" # remove extension
       [[ ! -e "$out/bin/$tName" ]] || continue
-      ln -sv "$(realpath $out/share/texmf/scripts/$s)" "$out/bin/$tName" # wrapped below
+      ln -sv "$(realpath "$TEXMFDIST/scripts/$s")" "$out/bin/$tName" # wrapped below
```
This line could be moved to `${texmf}` now?

xworld21 replied:
I suppose I'd prefer to leave the `bin` folder responsibility to the derivation that also does the shebangs and the wrapping. At least for now, because I have something else in mind for the script (in short: why not generate & patch the `bin` link right when creating the package, with the appropriate care not to destroy the fixed outputs and so on – and combine does wrapping only).

xworld21 replied:
Note for the future: generate `bin` script links early on. Best scenario, generate the `patchSheBangs`'d scripts right when building the package itself, so that `texlive.combine` is only in charge of the wrappers.

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -283,7 +292,7 @@ in (buildEnv {
   # and render the build unreproducible by their embedded timestamps.
   ''
-    find $TEXMFSYSVAR/web2c -name '*.log' -delete
+    find "$TEXMFSYSVAR"/web2c -name '*.log' -delete
```
Not relevant for this PR

xworld21 replied:
Yeah, that was just linting on my part, I can take it out.

xworld21 replied:
I'll keep the lint – we should always quote strings!

veprbl commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -262,16 +281,6 @@ in (buildEnv {
-    for d in {man,info}; do
-      [[ -e "$out/share/texmf/doc/$d" ]] || continue;
-      mkdir -p "$out/share/$d"
-      ln -s -t "$out/share/$d" "$out/share/texmf/doc/$d"/*
```
I'm confused by this bit. Is this not taken care of anymore?

veprbl replied:
Nvm. I see, that's done in `${texmf}`.

xworld21 replied:
This happens in `texmf` already. I wouldn't do that here, precisely because it would interfere with the man pages in `bin.out`.

xworld21 replied:
Oh, you had replied while I was typing!

veprbl reviewed (commented):
Looks okay for the most part. I do have some questions, though.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -38,16 +38,37 @@ let
+    name = "${name}-texmf";
+
+    paths = pkgList.nonbin;
+    extraPrefix = "/share/texmf";
```
The ideal solution would be of course to have no `extraPrefix`, and no rebuilding of the fixed output bits (then we could just point `TEXMFDIST` to `${texmf}`). However, we also need to generate `/share/man`, `/share/info` in the `texmf` derivation, hence the approach I have taken here.

veprbl replied:
I don't see the issue with rebuilding the fixed output bits. We'll need to update hashes. Backwards compatibility with custom packages can be provided by marking updated packages with an attribute like `__prefixIncluded = true`.

Also, it's not clear if those man/info directories are even needed. That piece of code was in the original version
https://github.com/NixOS/nixpkgs/commit/22fd6e1e4c52f8ab0651409812f9c53a59c1077c#diff-ab5e9b3939fcfd8e5a6e15cab07662787121eb16d7a7f92760d48941fbc90802R186-R192
But there is no comment as to what requires it.

xworld21 replied:
> if those man/info directories are even needed

AFAIU, the point is that some `doc` packages contain man pages, and either you override MANPATH, or add a symlink to make those available. I suppose alternatively, one can use texdoc to look them up (and we fixed texdoc some time ago).

I suggest to keep this conversation unresolved, while I figure out what Debian does. If they put the docs in the MANPATH, I say we do this in nixpkgs too.

xworld21 replied:
Ok, so Debian does move the man pages in MANPATH, which means we should too. I have also found an even better `buildEnv` to have man and info pages linked with less `extraPrefix` nonsense. I'll push after all the tests are happy.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -57,54 +78,40 @@ in (buildEnv {
     TEXMFSYSCONFIG="$out/share/texmf-config"
     TEXMFSYSVAR="$out/share/texmf-var"
-    export PERL5LIB="$out/share/texmf/scripts/texlive:${bin.core.out}/share/texmf-dist/scripts/texlive"
+    export TEXMFCNF="$TEXMFSYSVAR/web2c"
```
I am fairly sure it's ok. Kpathsea will read all `texmf.cnf` files it can find, but apply the settings in order of priority, with `TEXMFCNF` coming first. So the parts we change will apply, and if we forget to copy some file, it will still be read from `TEXMFDIST`.

(To be sure, I checked the documentation right now: https://www.tug.org/texinfohtml/kpathsea.html#Config-files)

xworld21 replied:
PS: on second though, the behaviour of Kpathsea means we can simply write the relevant variables in a new empty `texmf.cnf`, instead of using sed and regular expressions. But I was trying to keep this PR on point – this is rather something to try and test carefully in the future.

xworld21 replied:
Note for the future: avoid creating an entire copy of `texmf.cnf{.lua}` and just override the variables that need overriding. I'd prefer not to do this here as I'd prefer to test it separately, after the split is proven to work correctly.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -57,54 +78,40 @@ in (buildEnv {
     # tried inspiration from install-tl, sub do_texmf_cnf
   ''
-    if [ -e "$TEXMFCNF/texmfcnf.lua" ]; then
+    mkdir -p "$TEXMFCNF"
+    if [ -e "$TEXMFDIST/web2c/texmfcnf.lua" ]; then
```
As above, if we don't copy something over, the original will still be read from `TEXMFDIST`.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -132,18 +139,30 @@ in (buildEnv {
+        # enable the requested formats
+        + lib.concatMapStrings activateFormats formatPNames
+        # clean up formats that have been disabled twice
+        + "s/^#! #! /#! /;\n"
```
So this is where it gets tricky, because both files in `TEXMFDIST` and `TEXMFSYSVAR` are read by `fmutil`. The old combine could simply delete the missing formats; now, we cannot delete them from `TEXMFDIST`, thus we must *disable* them by prepending `#!` (as per documentation [1]).

This required some fiddling with `fmtutil.sed` about adding `#!`, then removing it for the packages that are being installed, and finally cleaning up the stray double `#!`'s caused by formats that are already disabled in the CTAN file (to be fair, the clean up is probably redundant). I hope the code is not too obscure.

[1] Section "Disabling formats" https://manpages.debian.org/experimental/texlive-base/fmtutil.1.en.html

xworld21 replied:
Added an extra documentation line referring to `man fmtutil`.

xworld21 committed:
texlive.combine: split static $TEXMFDIST into a separate derivation

xworld21 committed:
texlive.combine: use separate derivation for info and man pages

xworld21 commented:
@veprbl I have fixed two glaring typos (`$TEXMFDIST/share/texmf` instead of just `$TEXMFDIST` – glad I caught it before breaking LuaTeX!) and added an extra commit that avoids the ugly `/share/texmf`, at the price of an extra derivation. It should be good to merge, I think.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1901

SuperSandro2000 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -38,16 +38,44 @@ let
+      perl -I "${bin.core.out}/share/texmf-dist/scripts/texlive" \
+        -- "$out/scripts/texlive/mktexlsr.pl" --sort "$out"
+    '';
+  }).overrideAttrs (_: { allowSubstitutes = true; });
```
Does buildEnv really not take allowSubstitutes? If so we should fix this at buildEnv level.

xworld21 replied:
That would be great. Right now `buildEnv` only accepts a fixed list of attributes. We could add e.g. `extraArgs` to `buildEnv`?

SuperSandro2000 replied:
or better we accept the standard ones mkDerivation does and that make sense, too. extraArgs is usually only an escape hatch.

xworld21 replied:
@SuperSandro2000 do you mind if I start a separate `buildEnv` PR for this? In the meanwhile, if you don't spot anything suspicious and @veprbl is also happy, we merge this one.

I am mindful of the TeX Live 2022 final release approaching and it would be best not to have pending TeX Live infrastructure changes while that happens. I assume that a change to `buildEnv` will take some time, even if tiny, given how central it is to Nixpkgs.

SuperSandro2000 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -38,16 +38,44 @@ let
+  }).overrideAttrs (_: { allowSubstitutes = true; });
+
+  # expose info and man pages in usual /share/{info,man} location
+  doc = (buildEnv {
+    name = "${name}-doc";
+
+    paths = [ (texmf.outPath + "/doc") ];
+    extraPrefix = "/share";
+
+    pathsToLink = [
+      "/info"
+      "/man"
+    ];
+  });
```
```suggestion
  doc = buildEnv {
    name = "${name}-doc";

    paths = [ (texmf.outPath + "/doc") ];
    extraPrefix = "/share";

    pathsToLink = [
      "/info"
      "/man"
    ];
  };
```

veprbl commented:
@xworld21 This fell off my radar. Let us know when it's good to merge.

veprbl commented:
@GrahamcOfBorg build tests.texlive

wentasah commented:
The second commit breaks `tlType == "doc"` packages. If this is intentional, it should be mentioned in 23.05 release notes and [the manual](https://nixos.org/manual/nixpkgs/unstable/#sec-language-texlive-user-guide) should be updated.

Specifically, I mean that the following commands printed `OK` before this PR but now, `ERROR` is printed.
```sh
nix-build --expr 'with import ./. {}; 
(texlive.combine { 
   inherit (texlive) scheme-basic; 
   pkgFilter = (pkg: pkg.tlType == "run" || pkg.tlType == "bin" || pkg.pname == "core" || pkg.tlType == "doc");
})'
[ -e result/share/texmf/doc/latex/hyperref/hyperref.pdf ] && echo OK || echo ERROR
```
These commands should be run from the root of `nixpkgs` directory. The difference from the default configuration is addition of `pkg.tlType == "doc"` to `pkgFilter`.

xworld21 commented:
> Specifically, I mean that the following commands printed `OK` before this PR but now, `ERROR` is printed.

I see. The 'correct' way of looking up documentation is via texdoc, which searches all the appropriate directories (TEXMFDIST and the likes). The fact that we only have a single TEXMFDIST, and that it used to be exposed under `$out/share/texmf`, is part historical accident and part implementation detail.

Still, this is a breaking change for some people, so I am more than happy to reintroduce the symlink `$out/share/texmf` and maintain backwards compatibility. Will PR in a moment.
